### DISCUSSION TITLE
Добавить кнопки в сообщение телеграм

### DISCUSSION THREAD

**Discussion by erpic-ru** [totumonline/totum-ru-issues-and-discussions#218] [Как это сделать?] [answered]
Пытаюсь добавить inline кнопки в телеграм-сообщение бота для обратной связи. Типа таких:
<img width="604" alt="Снимок экрана 2022-09-15 в 13 46 45" src="https://user-images.githubusercontent.com/91674114/190405033-4317adfe-ee3b-4f14-b647-c91fb9867189.png">

Запрос примитивный:
```
{
  "text": "Заявка №1",
  "chat_id": "chat_id",
  "reply_markup": {
    "inline_keyboard": [
      [
        {
          "text": "Согласовать",
          "callback_data": "https://ya.ru"
        },
        {
          "text": "Отказать",
          "callback_data": "https://google.com"
        }
      ]
    ]
  }
}

```
При отправке из Postman все отлично работает и кнопки показываются.
 
Крутил/вертел этот код. Настроил Remotes, чтобы запрос проверить, но так и не получилось сделать. 
Он отправляет сообщение, но без этих кнопок. Подскажите, пожалуйста, что я делаю не так? Код в Тотуме:

```
=:getFromScript(uri: $link; posts: $posts;)

//  Генерируем ссылку
~link: str`"https://api.telegram.org/bot" + "токен" + "/sendMessage"`

// Текст сообщения
~message: str`"<b>ID:</b>" ++ #result`
// Запрос
posts: rowCreate(field: "chat_id" = "chat_id"; field: "text" = $message; field: "reply_markup" = $markup;)

markup: rowlistCreate(field: "inline_keyboard" = $do)
do:listCreate(item: $val2)

val2:rowlistCreate(field: "text" = $buttons; field: "url" = $urls)

buttons: listCreate(item: "Согласовать"; item: "Отказать";)
urls: listCreate(item: "https://ya.ru"; item: "https://ya.ru";)
```

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Исследование этого вопроса показало следующее:

По умолчанию при использовании `getFromScript`, если не указан тип в заголовке данные отправляются как `application/x-www-form-urlencoded`.

Таким образом массив переданный в параметр `posts` передается на сервер телеграма. В этом массиве `chat_id` и `text` это строки (с ними все ок), а `reply_markup` это структура тотума и телеграм ее не может прочитать.

Оно в `reply_markup` ожидают текст в формате `json` и его надо сделать через `jsonCreate`.

Я еще немного по мелочам подшаманил и получился такой код:

```
=: getFromScript(uri: $link; post: "text" = $message; post: "chat_id" = "CHAT_ID_HERE"; post: "reply_markup" = $markup)

		//  Генерируем ссылку
		~link: str`"https://api.telegram.org/bot" + "BOT_KEY_HERE" + "/sendMessage"`

		// Текст сообщения
		~message: "Тестовое сообщение 3"

	// Маркапы
	markup: jsonCreate(data: $row)
		row: rowCreate(field: "inline_keyboard" = $list)
			list: listCreate(item: $row_list_buttons)
				row_list_buttons: rowlistCreate(field: "text" = $buttons; field: "callback_data" = $urls)
					~buttons: listCreate(item: "Согласовать"; item: "Отказать")
					~urls: listCreate(item: "https://ya.ru"; item: "https://ya.ru")
```

**Reply by erpic-ru** [reply]
Большое спасибо! Кнопки заработали.
Единственное информационное уточнение: callback_data и url - это разные сущности.
В первом случае - это отправка ботом данных без перехода и его потом можно отловить через вебхуки или getupdates, во втором - внешний переход по ссылке, например, на Remote с передачей параметров через get

**Reply by ttm-alexeypolunin** [reply]
Не знаю в какой момент появилось, но на текущий момент это так.

**Если используется callback_data, то приложене ожидает ответный сигнал!**

NOTE: After the user presses a callback button, Telegram clients will display a progress bar until you call answerCallbackQuery. It is, therefore, necessary to react by calling answerCallbackQuery even if no notification to the user is needed (e.g., without specifying any of the optional parameters).

https://core.telegram.org/bots/api#answercallbackquery
